src/Stratis.Bitcoin.Features.BlockStore/BlockStoreBehavior.cs

```diff
@@ -257,23 +257,35 @@ public Task AnnounceBlocks(List<uint256> blockHashesToAnnounce)
                     {
                         headers.Add(chainedBlock.Header);
                     }
-                    else if (chainBehavior.PendingTip.GetAncestor(chainedBlock.Height) != null)
-                    {
-                        continue;
-                    }
-                    else if (chainBehavior.PendingTip.GetAncestor(chainedBlock.Previous.Height) != null)
-                    {
-                        // Peer doesn't have this header but they do have the prior one.
-                        // Start sending headers.
-                        foundStartingHeader = true;
-                        headers.Add(chainedBlock.Header);
-                    }
                     else
                     {
-                        // Peer doesn't have this header or the prior one -- nothing will
-                        // connect, so bail out.
-                        revertToInv = true;
-                        break;
+                        ChainedBlock peersBlock = chainBehavior.PendingTip.GetAncestor(chainedBlock.Height);
```

Comment:
this code is non-trivial, i'd suggest adding comments to every branch and sub-branch of this execution flow

Reply:
I would suggest that we search by hash, but searching by height is much faster (now with skip list)

Reply:
>I would suggest that we search by hash, but searching by height is much faster (now with skip list)

We check the hash later anyway so in this case it's better to do a faster search by height imho.